### ISSUE TITLE
Make all structs that claim to be thread-safe dereference `this` at most once per entrypoint

### ISSUE THREAD

**Issue by AArnott** [dotnet/corefx#2119]
`ImmutableArray<T>` goes to great lengths to be thread-safe by [never dereferencing `this` more than once per public entrypoint][1]. Other structs in the Framework do *not* take this care (such as `DateTime` and even `System.Int32`), despite [claiming to be thread-safe in MSDN documentation][2].

Should we take the same care for the other structs? 

 [1]: https://github.com/dotnet/corefx/blob/1dc118cc46f88e4889a23190b450d89e565967fd/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray%601.cs#L18-L33
 [2]: https://msdn.microsoft.com/en-us/library/System.DateTime(v=vs.110).aspx

**Comment by sharwell** [comment]
1. To answer your question, *yes*. This is a non-breaking change that corrects a clear bug in the implementation semantics when compared to documented guarantees of the types.
2. All of your examples are actually part of coreclr, not corefx.

**Comment by AArnott** [comment]
Good point 2. I've opened [a coreclr issue](/dotnet/coreclr/issues/1167) to track that.

**Comment by karelz** [comment]
We will take changes in this direction as long as they are backwards compatible. Feel free to submit PRs per struct or bundle them.
